pkgs/development/haskell-modules/generic-builder.nix

```diff
@@ -330,18 +357,77 @@ stdenv.mkDerivation ({
       done
     ''}
 
     ${optionalString enableSeparateDocOutput ''
-    for x in ${docdir "$doc"}/html/src/*.html; do
-      remove-references-to -t $out $x
+    # Remove references back to $out but also back to $lib if we have
+    # docs. $lib is needed as it stores path to haddock interfaces in the
+    # conf file which creates a cycle if docs refer back to library
+    # path.
+    mkdir -p ${docDir}
+
+    for x in ${docDir}/html/src/*.html; do
+      remove-references-to -t $out -t ${libDir} -t ${binDir} ${optionalString enableSeparateDataOutput "-t $data"} $x
     done
-    mkdir -p $doc
     ''}
-    ${optionalString enableSeparateDataOutput "mkdir -p $data"}
+
+    ${optionalString hasLibOutput ''
+    # Even if we don't have binary output for the package, things like
+    # Paths files will embed paths to bin/libexec directories in themselves
+    # which results in .lib <-> $out cyclic store reference. We
+    # therefore patch out the paths from separate library if we don't have
+    # separate bin output too.
+    #
+    # If we _do_ have separate bin and lib outputs, we may still be in
+    # trouble in case of shared executables: executable contains path to
+    # .lib, .lib contains path (through Paths) to .bin and we have a
+    # cycle.
+    #
+    # Lastly we have to deal with references from .lib back into
+    # $out/share if we're not splitting out data directory.
+    #
+    # It may happen that we have hasLibOutput set but the library
+    # directory was not created: this happens in the case that library
+    # section is not exposing any modules. See "fail" package for an
+    # example where no modules are exposed for GHC >= 8.0.
+    if [ -d ${libDir} ]; then
+      find ${libDir} -type f -exec \
+        remove-references-to -t ${binDir} -t ${libexecDir} "{}" \;
+    fi
+    ''}
+
+    ${optionalString (hasLibOutput && ! enableSeparateDocOutput) ''
+    # If we don't have separate docs, we have to patch out the ref to
+    # docs in package conf. This will likely break Haddock
+    # cross-package links but is necessary to break store cycle…
+    find ${libDir}/ -type f -name '*.conf' -exec \
+      remove-references-to -t ${docDir} "{}" \;
+    ''}
+
+    ${optionalString (hasLibOutput && ! enableSeparateDataOutput) ''
+    # Just like for doc output path in $out potentially landing in
+    # *.conf, we have to also remove the data directory so that it
+    # doesn't appear under data-dir field creating a cycle.
+    find ${libDir}/ -type f -exec echo Removing ${dataDir} refs from "{}" \;
+    find ${libDir}/ -type f -name '*.conf' -exec \
+      remove-references-to -t ${dataDir} "{}" \;
+    ''}
+
+    ${optionalString enableSeparateDataOutput "mkdir -p ${dataDir}"}
+    ${optionalString enableSeparateBinOutput "mkdir -p ${binDir} ${libexecDir}"}
+    ${optionalString enableSeparateEtcOutput "mkdir -p ${etcDir}"}
+
+    ${optionalString (stdenv.isDarwin && enableSeparateBinOutput && ! hasLibOutput) ''
```

Comment:
Is there a particular reason to do this on darwin only?

`happy` seems to [need](https://hydra.nixos.org/build/66728685/nixlog/1) something like this on linux.

Why not something like:

    configureFlags = drv.configureFlags or [] ++ optional (enableSeparateBinOutput && ! hasLibOutput) "--libdir=/var/empty";

That way we don't have to `remove-references-to` and it may (?) fail earlier. I'd rather replace it with an empty directory in the nix store though.